### ISSUE TITLE
Module version calculation ignores version of a build dependency

### ISSUE THREAD

**Issue by zyulyaev** [garden-io/garden#2566]
### Current Behavior

I have following configuration:
```
kind: Module
type: exec
name: web-build
local: true
build:
  command:
    - npm run build
exclude:
  - .next
  - node_modules
  - helm
env: ${var.env}

---

kind: Module
type: exec
name: web-copy
build:
  command:
    - cp ${local.projectPath}/package-lock.json . && node ${local.projectPath}/build-scripts/generate-package-json.js -C ${local.projectPath} @mm/web > package.json
include: []

---

kind: Module
type: container
name: web-image
dockerfile: Dockerfile
include:
  - Dockerfile
  - public
  - next.config.production.js
build:
  dependencies:
    - name: web-build
      copy:
        - source: .next
          target: .
    - name: web-copy
      copy:
        - source: package.json
          target: .
        - source: package-lock.json
          target: .
```

So the `web-image` module has a build dependency on the `web-build` module. The documentation states:
> Each version also factors in the versions of every dependency (both build and runtime dependencies, as is applicable for each case). This means that anytime a version of something that is depended upon changes, every dependant's version also changes.

So I expect the `web-image` module version to change if `web-build` version changes. However, this is not the case. Here are the logs from our CI for two builds of different branches:

### Build 1
```
Deploy 
ℹ providers                 → Getting status...
   ℹ Fetching kubectl...
   ✔ Fetched kubectl
✔ providers                 → Getting status... → Done
ℹ graphql-api-build         → Getting build status for v-f8c972dc3c...
ℹ web-build                 → Getting build status for v-116e639e3f...
ℹ graphql-api-copy          → Getting build status for v-3b8f8884e7...
ℹ web-copy                  → Getting build status for v-ab7385114f...
ℹ graphql-api-build         → Building version v-f8c972dc3c...
ℹ web-build                 → Building version v-116e639e3f...
ℹ Fetching helm...
ℹ graphql-api-copy          → Building version v-3b8f8884e7...
ℹ web-copy                  → Building version v-ab7385114f...
✔ Fetched helm
✔ graphql-api-copy          → Building version v-3b8f8884e7... → Done (took 4.7 sec)
✔ web-copy                  → Building version v-ab7385114f... → Done (took 4.7 sec)
✔ graphql-api-build         → Building version v-f8c972dc3c... → Done (took 123.7 sec)
ℹ graphql-api-image         → Getting build status for v-6235d8ca66...
-> Deploying garden-util service in mochary-method-preview-feature-m-51cpkh namespace (was outdated)
ℹ garden-util               → Waiting for resources to be ready...
ℹ garden-util               → Resources ready
-> Deploying garden-util service in mochary-method-preview-feature-m-51cpkh namespace (was outdated) → Done!
✔ graphql-api-image         → Getting build status for v-6235d8ca66... → Already built
ℹ graphql-api               → Getting build status for v-8822e5be66...
ℹ graphql-api               → Building version v-8822e5be66...
✔ graphql-api               → Building version v-8822e5be66... → Done (took 1.6 sec)
ℹ graphql-api               → Deploying version v-b3026d4214...
   ℹ graphql-api               → Waiting for resources to be ready...
   ℹ graphql-api               → Resources ready
✔ graphql-api               → Deploying version v-b3026d4214... → Done (took 56.2 sec)
✔ web-build                 → Building version v-116e639e3f... → Done (took 924.8 sec)
ℹ web-image                 → Getting build status for v-6ed54d2690...
-> Deploying garden-util service in mochary-method-preview-feature-m-51cpkh namespace (was outdated)
ℹ garden-util               → Waiting for resources to be ready...
ℹ garden-util               → Resources ready
-> Deploying garden-util service in mochary-method-preview-feature-m-51cpkh namespace (was outdated) → Done!
ℹ web-image                 → Building version v-6ed54d2690...
-> Deploying garden-util service in mochary-method-preview-feature-m-51cpkh namespace (was outdated)
ℹ garden-util               → Waiting for resources to be ready...
ℹ garden-util               → Resources ready
-> Deploying garden-util service in mochary-method-preview-feature-m-51cpkh namespace (was outdated) → Done!
ℹ web-image                 → Syncing sources to cluster...
ℹ web-image                 → Building image web-image:v-6ed54d2690...
✔ web-image                 → Building image web-image:v-6ed54d2690... → Done (took 383 sec)
ℹ web                       → Getting build status for v-faa46f4cc8...
ℹ web                       → Building version v-faa46f4cc8...
✔ web                       → Building version v-faa46f4cc8... → Done (took 1 sec)
ℹ web                       → Deploying version v-c2c47e8b3a...
   ℹ web                       → Waiting for resources to be ready...
   ℹ web                       → Resources ready
✔ web                       → Deploying version v-c2c47e8b3a... → Done (took 60.5 sec)
Done! 
```

### Build 2

```
Deploy 
ℹ providers                 → Getting status...
   ℹ Fetching kubectl...
   ✔ Fetched kubectl
✔ providers                 → Getting status... → Done
ℹ graphql-api-build         → Getting build status for v-f8c972dc3c...
ℹ web-build                 → Getting build status for v-59c7d6fa0f...
ℹ graphql-api-copy          → Getting build status for v-3b8f8884e7...
ℹ web-copy                  → Getting build status for v-ab7385114f...
ℹ graphql-api-build         → Building version v-f8c972dc3c...
ℹ web-build                 → Building version v-59c7d6fa0f...
ℹ Fetching helm...
ℹ graphql-api-copy          → Building version v-3b8f8884e7...
ℹ web-copy                  → Building version v-ab7385114f...
✔ graphql-api-copy          → Building version v-3b8f8884e7... → Done (took 5.6 sec)
✔ web-copy                  → Building version v-ab7385114f... → Done (took 5.6 sec)
✔ Fetched helm
✔ graphql-api-build         → Building version v-f8c972dc3c... → Done (took 133 sec)
ℹ graphql-api-image         → Getting build status for v-6235d8ca66...
-> Updating Docker auth secret in namespace mochary-method-preview-staging-j5e8oc
-> Deploying garden-util service in mochary-method-preview-staging-j5e8oc namespace (was missing)
ℹ garden-util               → Waiting for resources to be ready...
ℹ garden-util               → Deployment/garden-util: Pulling image "gardendev/k8s-util:0.5.1"
ℹ garden-util               → Deployment/garden-util: Started container util
ℹ garden-util               → Resources ready
-> Deploying garden-util service in mochary-method-preview-staging-j5e8oc namespace (was missing) → Done!
✔ graphql-api-image         → Getting build status for v-6235d8ca66... → Already built
ℹ graphql-api               → Getting build status for v-2a876fe0e0...
ℹ graphql-api               → Building version v-2a876fe0e0...
✔ graphql-api               → Building version v-2a876fe0e0... → Done (took 1.6 sec)
ℹ graphql-api               → Deploying version v-3c9129181c...
   ℹ graphql-api               → Waiting for resources to be ready...
   ℹ graphql-api               → Resources ready
✔ graphql-api               → Deploying version v-3c9129181c... → Done (took 58.2 sec)
✔ web-build                 → Building version v-59c7d6fa0f... → Done (took 722.2 sec)
ℹ web-image                 → Getting build status for v-6ed54d2690...
-> Deploying garden-util service in mochary-method-preview-staging-j5e8oc namespace (was outdated)
ℹ garden-util               → Waiting for resources to be ready...
ℹ garden-util               → Resources ready
-> Deploying garden-util service in mochary-method-preview-staging-j5e8oc namespace (was outdated) → Done!
✔ web-image                 → Getting build status for v-6ed54d2690... → Already built
ℹ web                       → Getting build status for v-a8e0c8920e...
ℹ web                       → Building version v-a8e0c8920e...
✔ web                       → Building version v-a8e0c8920e... → Done (took 1 sec)
ℹ web                       → Deploying version v-a6d39c471d...
   ℹ web                       → Waiting for resources to be ready...
   ℹ web                       → Resources ready
✔ web                       → Deploying version v-a6d39c471d... → Done (took 61.3 sec)
Done! 
```

As you can see `web-build` has `v-116e639e3f` and `v-59c7d6fa0f` versions for build 1 and build 2 respectively. While the `web-image` module has `v-6ed54d2690` version for both builds.

### Expected behavior

A version of a module changes when its dependencies' versions change.

### Your environment

* **OS:** GitLab CI
* **How I'm running Kubernetes:** GKE autopilot
* **Garden version**: 0.12.25

**Comment by zyulyaev** [comment]
As far as I understand, the version calculation algorithm looks as following: `hash(this module's files, this module's configuration, dependencies' files)`. So it actually ignores dependencies' configuration such as build command arguments or environment variables which can and usually affect the build result.

**Comment by thsig** [comment]
Hi @zyulyaev—thanks for raising this question! We'll have a brief internal discussion on this as soon as possible and get back to you with an answer afterwards.

**Comment by thsig** [comment]
Hi @zyulyaev—we've had our internal discussion on this, and are working on a fix.

Note: The problem you highlighted is mostly limited to this specific case of using the `env`  field in `exec` modules that are used as build dependencies.

Thanks again for reporting this!

**Comment by thsig** [comment]
Closed by #2937.
